test/integration_test.go

```diff
@@ -0,0 +1,260 @@
+	time.Sleep(fetchInterval * time.Duration(fetchIterations))
+
+	// Collect metrics from the woodpecker instance while it is still running
+	metricsData := getWoodpeckerMetrics("http://localhost:1971")
+
+	// Stop the monitoring process
+	woodpecker.Stop()
+
+	// Return the stdout & the metrics data
+	return out.String(), metricsData, nil
+}
+
+// testServers creates & starts a number of CT test servers based on the
+// provided personalities. The servers are returned such that the caller can
+// cleanly shut them down when required.
+func testServers(personalities []cttestsrv.Personality) []*cttestsrv.IntegrationSrv {
+	var servers []*cttestsrv.IntegrationSrv
+	for _, p := range personalities {
+		logger := log.New(os.Stdout, fmt.Sprintf("ct-test-srv %q ", p.Addr), log.LstdFlags)
+		srv, err := cttestsrv.NewServer(p, logger)
+		if err != nil {
+			logger.Fatal(err)
+		}
+		servers = append(servers, srv)
+		srv.Run()
+	}
+
+	return servers
+}
+
+// defaultPersonalities returns hardcoded Personality data suitable for running
+// two test CT log servers, one on :4500, one on :4501.
+func defaultPersonalities() []cttestsrv.Personality {
+	return []cttestsrv.Personality{
+		cttestsrv.Personality{
+			Addr:    ":4500",
+			PrivKey: "MHcCAQEEIOCtGlGt/WT7471dOHdfBg43uJWJoZDkZAQjWfTitcVNoAoGCCqGSM49AwEHoUQDQgAEYggOxPnPkzKBIhTacSYoIfnSL2jPugcbUKx83vFMvk5gKAz/AGe87w20riuPwEGn229hKVbEKHFB61NIqNHC3Q==",
+			LatencySchedule: []float64{
+				0.1,
+				0.5,
+			},
+		},
+		cttestsrv.Personality{
+			Addr:    ":4501",
+			PrivKey: "MHcCAQEEIJSCFDYXt2xCIxv+G8BCzGdUsFIQDWEjxfJDfnn9JB5loAoGCCqGSM49AwEHoUQDQgAEKtnFevaXV/kB8dmhCNZHmxKVLcHX1plaAsY9LrKilhYxdmQZiu36LvAvosTsqMVqRK9a96nC8VaxAdaHUbM8EA==",
+			LatencySchedule: []float64{
+				0.5,
+				0.1,
+			},
+		},
+	}
+}
+
+// TestFetchSTHSuccess tests that a ct-woodpecker instance correctly checks the
+// STH of mock CT servers configured for success.
+func TestFetchSTHSuccess(t *testing.T) {
+	// root is an encoded SHA256 hash that we can jam into mock STHs
+	var root ct.SHA256Hash
+	_ = root.FromBase64String("ZVWlmKuutzCIAIjNuVW0kYrk69eqWbNtLX86CBMVneg=")
+	now := time.Now()
+
+	// Create and start some CT test servers with the default personalities
+	testServers := testServers(defaultPersonalities())
+
+	// Generate a mock STH for each of the CT test servers
+	mockSTHs := make([]*ct.SignedTreeHead, len(testServers))
+	for i, srv := range testServers {
+		// Offset each log's fake STH into the past by 1 hour + the log index
+		offset := -time.Duration(i+1) * time.Hour
+		ts := now.Add(offset).UnixNano() / int64(time.Millisecond)
+
+		// Generate a new mock STH for the log at the calculated timestamp offset
+		mockSTHs[i] = &ct.SignedTreeHead{
+			TreeSize:       0xC0FFEE,
+			SHA256RootHash: root,
+			Timestamp:      uint64(ts),
+		}
+		// Set the mock STH
+		srv.SetSTH(mockSTHs[i])
+	}
+
+	// Create a CT woodpecker configuration that fetches the STH of the two test logs every 5s
```

Comment:
Any reason to use 5s rather than say 100ms?

Reply:
No reason. Good suggestion!